### PR TITLE
Fix stacktrace unit test in Trilinos

### PR DESCRIPTION
**Pull request by dalg24**
Alternative to #2682

~~*disclaimer*:  I haven't actually tried :)~~

### CODE REVIEW

**Comment by masterleinad** [comment]
@ndellingwood It would good to know if this actually makes the `StackTrace` working when building `Kokkos` through `Trilinos`.

**Comment by ndellingwood** [comment]
@masterleinad I'll try and get to testing this out a bit later today, finishing a fix in kokkos-kernels exposed during the Trilinos PR testing with the release-candidate that didn't occur during my integration testing.

**Comment by masterleinad** [comment]
Thanks!

**Comment by dalg24** [comment]
Ping

**Comment by ndellingwood** [comment]
I haven't had time to get this, been having fun addressing other issues in Trilinos since the release. I'll try and test soon.

**Comment by ndellingwood** [comment]
@dalg24 @masterleinad 

I tried this in Trilinos doing a source override for kokkos and the test failed to run

```
[ndellin@kokkos-dev-2 KokkosOverride-serial]$ ctest -R KokkosCore_UnitTest_StackTraceTest
Test project /ascldap/users/ndellin/Trilinos/Build/KokkosOverride-serial
    Start 5: KokkosCore_UnitTest_StackTraceTest
1/1 Test #5: KokkosCore_UnitTest_StackTraceTest ...***Not Run (Disabled)   0.00 sec
No tests were found!!!
```

Relevant details:
1. I merged master branch into this pr to ensure it was compatible with Trilinos (I confirmed I was using the changes to ` core/unit_test/CMakeLists.txt` made in this PR)
2. Kokkos develop branch breaks Trilinos at configuration when trying to do a source override, I believe due to merge of PR #2688, preventing testing of Trilinos with the kokkos' repo develop branch, I will file an issue with the details
3. Tested on kokkos-dev-2 with configuration:

```bash
module load sems-cmake/3.12.2 sems-gcc/7.2.0 sems-git

rm -rf CMake*

export TRILINOS_PATH=${PWD}/../..

INSTALL_LOCATION=${PWD}/install
 
cmake \
-D CMAKE_INSTALL_PREFIX:PATH="${INSTALL_LOCATION}" \
-D CMAKE_BUILD_TYPE:STRING="RELEASE" \
-D KOKKOS_ARCH="SKX" \
-D Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=OFF \
-D Trilinos_ENABLE_TESTS:BOOL=OFF \
-D Trilinos_ENABLE_EXAMPLES:BOOL=OFF \
\
-D Trilinos_ENABLE_Kokkos:BOOL=ON \
 -D Kokkos_ENABLE_SERIAL:BOOL=ON \
 -D Kokkos_ENABLE_TESTS:BOOL=ON \
 -DKokkos_SOURCE_DIR_OVERRIDE:STRING=kokkos \
 -DKokkosKernels_SOURCE_DIR_OVERRIDE:STRING=kokkos-kernels \
${TRILINOS_PATH}
```

Note, the source override requires symbolic links to the kokkos and kokkos-kernels repos be added to the Trilinos root directory.

**Comment by dalg24** [comment]
`***Not Run (Disabled)` ??
Why does it say "disabled"?  I don't remember this being there when the executable is not found (because it was not build or because the name/path is wrong)

**Comment by ndellingwood** [comment]
> Why does it say "disabled"? I don't remember this being there when the executable is not found (because it was not build or because the name/path is wrong)

@dalg24 I'm not certain, are the reproducer notes I provided sufficient to investigate? I wanted to get some feedback to you but I don't have much time to spare digging into this at the moment.

**Comment by dalg24** [comment]
I closed this by mistake

**Comment by dalg24** [comment]
@ndellingwood would you please look into why the unit test was showing as "disabled" on the Trilinos side?

**Comment by ndellingwood** [comment]
> @ndellingwood would you please look into why the unit test was showing as "disabled" on the Trilinos side?

Yes, I have it in my queue, should get to it later today or tomorrow.

File: core/unit_test/CMakeLists.txt [hunk 1]
```diff
@@ -314,25 +314,13 @@ KOKKOS_ADD_TEST_EXECUTABLE(
 )
 # We need -rdynamic on GNU platforms for the stacktrace functionality
 # to work correctly with shared libraries
-if(NOT KOKKOS_HAS_TRILINOS)
-SET_PROPERTY(TARGET StackTraceTestExec PROPERTY ENABLE_EXPORTS 1)
+SET_PROPERTY(TARGET ${PACKAGE_NAME}_StackTraceTestExec PROPERTY ENABLE_EXPORTS ON)
 
-KOKKOS_ADD_TEST( NAME UnitTest_StackTraceTest_normal
-                 EXE  StackTraceTestExec
+KOKKOS_ADD_TEST( NAME UnitTest_StackTraceTest
```
Comment (ndellingwood) [inline core/unit_test/CMakeLists.txt]: @dalg24 if I replace lines 319-22

```
KOKKOS_ADD_TEST( NAME UnitTest_StackTraceTest
                 EXE  ${PACKAGE_NAME}_StackTraceTestExec
                 FAIL_REGULAR_EXPRESSION "FAILED"
              )
```

with this:
```
KOKKOS_ADD_ADVANCED_TEST( UnitTest_StackTraceTest
                 TEST_0
                  EXEC  StackTraceTestExec
                  NUM_MPI_PROCS 1
                  FAIL_REGULAR_EXPRESSION "FAILED"
               )
```

then the tests build and pass in Trilinos:

```
...
      Start  5: KokkosCore_UnitTest_StackTraceTest
 1/24 Test  #5: KokkosCore_UnitTest_StackTraceTest ...............   Passed    0.06 sec
...
```

Comment (ndellingwood) [reply core/unit_test/CMakeLists.txt]: This solution doesn't work with kokkos standalone, in that this removes the stacktrace test for standalone

https://github.com/kokkos/kokkos/blob/586d07ba08e2bb9aae6fe644cf7ac16853dd32bd/cmake/fake_tribits.cmake#L145-L151

`KOKKOS_ADD_ADVANCED_TEST` has no implementation outside of Trilinos usage (true of master used by Trilinos and VOTD develop)

Comment (dalg24) [reply core/unit_test/CMakeLists.txt]: > if I replace lines 319-22

Why did you change these lines in the first place?  What was wrong with them?

Comment (dalg24) [reply core/unit_test/CMakeLists.txt]: @jjwilke Would you please look into this

Comment (jjwilke) [reply core/unit_test/CMakeLists.txt]: Look into filling in the TODO? Basically you want an implementation of ADVANCED_TEST for Kokkos standalone?

**Comment by ndellingwood** [comment]
@jjwilke a couple separate issues:
1. The parsing of arguments (or usage) of `KOKKOS_ADD_TEST` before handoff to the pared TriBITs call  seems to be an problematic https://github.com/kokkos/kokkos/blob/4af1f35f15000ab9881ea964868a2b2f84c27199/cmake/fake_tribits.cmake#L88-L107 
If I configure Trilinos using Kokkos from this PR (modulo changes for compatibility with master branch) then the generated `CTestTestfile.cmake` file shows the following:

```
add_test(KokkosCore_UnitTest_StackTraceTest "UnitTest_StackTraceTest" "FALSE" "ADDED_TEST_NAME")
set_tests_properties(KokkosCore_UnitTest_StackTraceTest PROPERTIES  DISABLED "ON" FAIL_REGULAR_EXPRESSION "FAILED;FAILED" LABELS "Kokkos" PROCESSORS "/ascldap/users/ndellin/Trilinos/Build/KokkosOnly/kokkos/core/unit_test/KokkosCore_KokkosCore_StackTraceTestExec.exe" REQUIRED_FILES "UnitTest_StackTraceTest")
```

The parse args are incorrect; for reference here is what a test without issues looks like, where the test was added via `KOKKOS_ADD_TEST_EXECUTABLE` in the CMakeLists.txt

```
add_test(KokkosCore_UnitTest_DefaultInit_1 "/ascldap/users/ndellin/Trilinos/Build/KokkosOnly/kokkos/core/unit_test/KokkosCore_UnitTest_DefaultInit_1.exe")
set_tests_properties(KokkosCore_UnitTest_DefaultInit_1 PROPERTIES  FAIL_REGULAR_EXPRESSION "  FAILED  " LABELS "Kokkos" PROCESSORS "1" REQUIRED_FILES "/ascldap/users/ndellin/Trilinos/Build/KokkosOnly/kokkos/core/unit_test/KokkosCore_UnitTest_DefaultInit_1.exe")
```

If the problem is usage (missing TriBITs type args), can KOKKOS_ADD_TEST be updated to internally supply TriBITs-specific args instead of relying on user-supplied calls from CMakeLists.txt?

2. Filling in the TODO for the KOKKOS_ADD_ADVANCED_TEST for standalone would be helpful (and possibly necessary for this unit test to function with Trilinos), and will be necessary to enable the `push_finalize_hook_terminate` test in standalone Kokkos.

Thanks!

**Comment by jjwilke** [comment]
I believe the weirdness with StackTrace tests have been fixed (with some extra sanity checks). @dalg24 @ndellingwood

**Comment by dalg24** [comment]
CI does not pass neither on Jenkins nor Travis :/

**Comment by dalg24** [comment]
@ndellingwood please review

**Comment by ndellingwood** [comment]
I just verified @jjwilke recent commit addressed the issue, allowing the test to run and pass in standalone kokkos and trilinos.

**Comment by crtrott** [comment]
If you all are happy with this merge it in please @jjwilke @dalg24 @ndellingwood

**Pull request merged**